Ready for review (#2)
lsst-dm/dmtn-178: howto.tex

\section{Introduction}

> This is intended to be a short guide to help the installation and  execution of docsteady.

Reviewer: ```suggestion
This is intended to be a short guide to the installation and use of docsteady.
```

See \cite{DMTN-140} for a high level overview of the process and the logic implemented here.

\section{Installation}\label{sec:install}

Create a conda environment based on the \textbf{docsteady} conda package:

\texttt{conda create --name docsteady-env docsteady -c lsst-dm}

> Ensure that the conda configuration file  \texttt{.condarc} does not include the conda-forge channel.

Reviewer: I think this is best expressed using `-c lsst-dm -c defaults --override-channels` if that's what you want.

Author: OK. 
This implies that we do a new release of docsteady and we make it available in lsst-dm conda channel

To use docsteady, activate the environment as follows:

\texttt{conda activate docsteady-env}

This environment will provide all dependencies that are required to run docsteady.

It is recommended to use the provided conda environment also for development activities, see section \ref{sec:development}.

\section{Execution}

> It is recommended to provide the credentials seting up the following environment variables:

Reviewer: ```suggestion
It is recommended to provide credentials by setting up the following environment variables:
```

\begin{itemize}
\item \texttt{export JIRA\_USER=<jira-username>}
\item \texttt{export JIRA\_PASSWORD=<password>}
\end{itemize}

otherwise, it is required to specify them from the command line using the \texttt{--username} and \texttt{--password} options.
In case credential options are omitted and no environment variables are defined, username and password will be prompted interactively.

Personal Jira credentials can be used. 
For CI purposes, a general set of credentials are available, as specified in section \ref{sec:auth}.

In order to execute any of the docsteady commands described in the following subsections, a conda environment providing docsteady shall be activated, as described in section \ref{sec:install}.

Use \texttt{--help} to get the list of available options for each of the docsteady commands.

> In addition, existing documents generated using docsteady, have a \texttt{.docugen} file with the command to be executed in CI, which can be used as an example.

Reviewer: ```suggestion
In addition, existing documents generated using docsteady have a \texttt{.docugen} file with the command to be executed in CI, which can be used as an example.
```

\subsection{Test Specification Generation}

> Test specifications are extracted using REST API.

Reviewer: ```suggestion
Test specifications are extracted using the Jira REST API.
```

All tests cases included in a TM4J folder, including subfolders, are rendered in the same extraction.

> The folder organization in Jira should correspond to the major subsystem components, and modeled in MagicDraw.

Reviewer: ```suggestion
The folder organization in Jira should correspond to the major subsystem components and the MagicDraw model.
```

The syntax to extract a test specification is the following:

\texttt{docsteady generate-spec "</tm4j\_folder>" jira\_docugen.tex}

where \texttt{</tm4j\_folder>} shall be replaced by the exact folder where the test cases are defined in the Jira \texttt{Test Cases} tab.
For example, the command to extract the DM Acceptance test specification, LDM-639, is the following:

\texttt{docsteady generate-spec "/Data Management/Acceptance|LDM-639" [jira\_docugen.tex]}

Note that:
\begin{itemize}

> \item the output file \textit{jira\_docugen.tex} is optional in the execution of docsteady, but required in this context in order to include the extracted information in a \LaTeX~document (i.e. LDM-639.tex). If omitted, the docsteady  output will just be printed in the terminal;

Reviewer: ```suggestion
\item the output file \textit{jira\_docugen.tex} is optional in the execution of docsteady but required in this context in order to include the extracted information in a \LaTeX~document (i.e. LDM-639.tex). If omitted, the docsteady output will just be printed in the terminal;
```

> \item the folder name in Jira includes at the end the test specification document handle. This is a best practice to use when organizing test cases in Jira, it helps to orientate in the folder structure;

Reviewer: ```suggestion
\item the folder name in Jira includes at the end the test specification document handle. This is a best practice to use when organizing test cases in Jira since it helps to orient the user in the folder structure;
```

> \item an appendix with the traceability to the requirements is produced in the file \textit{jira\_docugen.appendix.tex}, to be included in the test specification tex file. 

Reviewer: ```suggestion
\item an appendix with the traceability to the requirements is produced in the file \textit{jira\_docugen.appendix.tex}, to be included in the test specification TeX file. 
```

\end{itemize}

See LDM-639 Git repository as an example.

\subsection{Test Plan and Report Generation}

\textbf{Important}: before extracting a test plan and report using docsteady,
the corresponding document handle has to be added in the \texttt{Document ID} field in the Jira test plan object.
This ensures that the Verification Control Document will include this information.

The following command extracts a Test Plan and Report using Jira REST API:

\texttt{docsteady generate-tpr <LVV-PXX> <file.tex> [--trace true]}

Where:

\begin{itemize}
\item \texttt{LVV-PXX} is the TM4J object that describes the test campaign, for example \texttt{LVV-P72};
\item \texttt{file.tex} is the Test Plan and Report tex file where the document will be rendered, for example \texttt{DMTR-231.tex};
\item \texttt{--trace true} (optional) generates an appendix with the traceability information.
\end{itemize}

Each TM4J test plan and related information in Jira is rendered in a different Test Plan and Report document,

> which filename usually corresponds also to the document handle in Docushare.

Reviewer: ```suggestion
the filename of which usually corresponds also to the document handle in DocuShare.
```

> The generated file can be built directly into the corresponding pdf, however additional files are required.

Reviewer: ```suggestion
The generated file can be built directly into the corresponding pdf, however additional files are required:
```

\begin{itemize}
\item Makefile
\item appendix.tex
\item history\_and\_info.tex
\end{itemize}

When creating the Git repository using \textbf{sqrbot-jr}, all the required files should already be present.
See \href{https://sqr-006.lsst.io/}{SQR-006} for more information regarding sqrbot-jr.

In case you want to generate a Test Plan and Report for a different subsystem, not DM, you can use the namespace global option:

\texttt{docsteady --namespace <NS> generate-tpr <LVV-PXX> <file.tex> [--trace true]}

Valid namespaces are:

\begin{itemize}
\item SE: system Enginering
\item DM: Data Management
\item T\&S: Telescope \& Site
\end{itemize}

See SCTR-14 or DMTR-231 Git repositories as an example.

\subsection{Verification Element Baseline Generation}

Verification Elements (VE) are Jira issues in the LVV Jira project, of type \texttt{Verification}.

> They are categorized in Components (DM, SITCOM, etc) and Sub-Components.

Reviewer: ```suggestion
They are categorized into Components (DM, SITCOM, etc) and Sub-Components.
```

A VE baseline document is extracted using REST API.
All VE associated with a Jira Component or Sub-Component, if specified, are rendered in the same extraction.

The syntax to extract a VE baseline information is the following:

\texttt{docsteady [--namespace <CMP>] baseline-ve [--subcomponent <SUBC>] jira\_docugen.tex [--details true]}

The information is saved in the specified \texttt{jira\_docugen.tex} file.
This file has to be included in a \LaTeX~document, where the corresponding context about the Component and Sub-Component is provided.

The \texttt{--namespace <CMP>} option identifies the Jira component from which to extract the information.
The parameter \texttt{CMP} shall correspond to the Rubin Observatory sub-systems.

> See the See subsection \ref{sec:components} for the complete list of components.

Reviewer: ```suggestion
See subsection \ref{sec:components} for the complete list of components.
```

A direct access to the Jira database implies also that the username and password to use are different since credentials to access the Jira web interface or the REST API are not enabled to access the database. They are two different authentication systems.
Therefore personal Jira credentials will not work with this docsteady command.

A special read-only user has been enabled in the Jira database, \textbf{jiraro}.
Section \ref{sec:auth} explains where to find the full credentials details.

For your convenience, the credentials can be specified in the following environment variables:

\begin{itemize}
\item \texttt{export JIRA\_VCD\_USER=jiraro}
\item \texttt{export JIRA\_VCD\_PASSWORD} (see section \ref{sec:auth})
\item \texttt{export JIRA\_DB=140.252.201.12}
\end{itemize}

otherwise, it is required to specify them from the command line using the options \texttt{--vcduser}, \texttt{--vcdpwd}, and \texttt{--jiradb}.
In case credential options are omitted and no environment variables are defined, they will be prompted interactively.
Note also that the Jira database IP address may change. Updated information are maintained in the vault specified in section \ref{sec:auth}.

> The following command extractsall VCD information regarding \textbf{DM} and generates the file \texttt{jira\_docugen.tex}:

Reviewer: ```suggestion
The following command extracts all VCD information regarding \textbf{DM} and generates the file \texttt{jira\_docugen.tex}:
```

\texttt{docsteady [--namespace <COM>] generate-vcd --sql True jira\_docugen.tex}

When no \texttt{--namespace if provided}, the DM component is selected by default.
The generated file \textbf{jira\_docugen.tex} is meant to be included in LDM-692.tex.

> In case you want to generate the VCD for a different LSST/Rubin Observatory subsystem VCD,

Reviewer: ```suggestion
In case you want to generate the VCD for a different LSST/Rubin Observatory subsystem,
```

just use the corresponding subsystem code configured in the Jira \textbf{component} field. 
See next subsection \ref{sec:components} for the complete list.

\subsection{Components - Sub-systems}\label{sec:components}

Follows the list of components configured for the Jira LVV project.
Each component corresponds to a Rubin Observatory Construction subsystem.

\begin{itemize}
\item \textbf{CAM}: Camera
\item \textbf{DM}: Data Management, the default component for all docsteady commands.
\item \textbf{EPO}: Education and Public Outreach
\item \textbf{OCS}: Observatory Control System
\item \textbf{PSE}: Project System Engineering, used for Commisioning (SitCom)
\item \textbf{T\&S}: Telescope and Site
\end{itemize}

> In case the subcomponent specified is "None", all VE without subcomponent will be extracted.

Reviewer: ```suggestion
In case the subcomponent specified is "None", all VE without subcomponents will be extracted.
```

\section{Development}
\label{sec:development}

> Despite docsteady is a pure python tool, it depends on \textbf{pandoc}, that is a \texttt{c++} compiled library available only as conda package.

Reviewer: ```suggestion
Despite docsteady being a pure python tool, it depends on \textbf{pandoc}, which is a \texttt{C++} compiled library available only as a conda package.
```

> It has been observed also, that any small change in the version of pandoc may lead to unexpected changes in the resulting \LaTeX~format.

Reviewer: ```suggestion
It has been observed that any small change in the version of pandoc may lead to unexpected changes in the resulting \LaTeX~format.
```

Therefore, in order to ensure the expected pandoc behavior, it is important to set-up the conda environment corresponding to the latest docsteady working version.
The environment set-up is explained in section \ref{sec:install}.

> The docsteady source code of available at \url{https://github.com/lsst-dm/docsteady}.

Reviewer: ```suggestion
The docsteady source code is available at \url{https://github.com/lsst-dm/docsteady}.
```

To test changes done locally in the source code, use the following procedure:

\begin{itemize}

> \item (if not available) create the environment as specified in section \ref{sec:install}

Reviewer: ```suggestion
\item (if not already available) create the environment as specified in section \ref{sec:install}
```

\item activate the environment: \texttt{conda activate docsteady-env}
\item clone docsteady repository and checkout a ticket branch
\item do your changes
\item install the updates in the docsteady-env environment: \texttt{python setup.py install}

> \item activate the same docsteady-env environment in a different terminal to test the new changed

Reviewer: ```suggestion
\item activate the same docsteady-env environment in a different terminal to test the new changes
```

\item once the changes are OK, commit them in the repository and open a PR for merging the branch to master
\end{itemize}

\section{Documentation Procedure}
\label{sec:docproc}

The autogeneration of documents may become very confusing if not done in a programmatic way.
Please consider the DM documentation approach as a guideline, summarized here.

\begin{itemize}
\item Create a document handle in DocuShare

> \item Use the document handle to create a repository in Github, using sqrbot-jr that will also create the corresponding landing page in lsst.io

Reviewer: ```suggestion
\item Use the document handle to create a repository in GitHub using sqrbot-jr, which will also create the corresponding landing page in lsst.io
```

> \item Configure the continuous integrations described in next section ~\ref{sec:ci}.

Reviewer: ```suggestion
\item Configure the continuous integrations described in the next section ~\ref{sec:ci}.
```

This service is at the moment behind the NCSA firewall, and therefore it requires a VPN to access and monitor the jobs.
Authentication to this Jenkins instance is managed via OAuth. See next sub-section \ref{sec:auth}.

The rendered documents will be available in the corresponding GitHub repositories and LSST The Docs landing pages.

The docugen jobs are created using the seeds script defined at the following location:

\url{https://github.com/docs-ci/docs-ci-seeds/blob/master/jobs/docugen_jobs.groovy}

There should be no need to change this script, unless problems are found in its logic, or improvements need to be implemented
The jobs are configured in the following YAML file:

\url{https://github.com/docs-ci/docs-ci-seeds/blob/master/etc/docugen.yaml}

The seeds job in Jenkins:

\url{https://lsst-docs-ci.ncsa.illinois.edu/jenkins/job/Service/job/seeding/}

will detect the changes in the YAML configuration file and create/update the the corresponding Jenkins docugen jobs.

\subsection{Authentication}\label{sec:auth}

The access to \url{https://lsst-docs-ci.ncsa.illinois.edu/jenkins} is granted adding users to the \textbf{docs-ci} GitHub organization, 

> managed by DM Architecture team.

Reviewer: ```suggestion
managed by the DM Architecture team.
```

> Two generic set of credentials to access Jira REST API and the Jira database have been defined.

Reviewer: ```suggestion
Two generic set of credentials to access the Jira REST API and the Jira database have been defined.
```
